File: taskmanager/src/main/java/com/example/alexeyglushkov/taskmanager/task/TaskManagerSnapshot.java
package com.example.alexeyglushkov.taskmanager.task;

import android.util.SparseArray;

import androidx.collection.SparseArrayCompat;

/**
 * Created by alexeyglushkov on 23.08.15.
 */
public interface TaskManagerSnapshot {

    void startSnapshotRecording(TaskManager taskManager);
    void stopSnapshotRecording();

    void addSnapshotListener(OnSnapshotChangedListener listener);
    void removeSnapshotListener(OnSnapshotChangedListener listener);

    int getLoadingTasksCount();
    int getWaitingTasksCount();
    int getMaxQueueSize();
    SparseArrayCompat<Float> getLoadingLimits();
    SparseArrayCompat<Integer> getUsedLoadingSpace();
    SparseArrayCompat<Integer> getWaitingTaskInfo();

    interface OnSnapshotChangedListener {
        void onSnapshotChanged(TaskManagerSnapshot snapshot);
    }
}

File: taskmanager/src/main/java/com/example/alexeyglushkov/taskmanager/task/TaskManagerSnapshot.kt
package com.example.alexeyglushkov.taskmanager.task

import android.util.SparseArray

import androidx.collection.SparseArrayCompat

/**
 * Created by alexeyglushkov on 23.08.15.
 */
interface TaskManagerSnapshot {
    val loadingTasksCount: Int
    val waitingTasksCount: Int
    val blockedTasksCount: Int
    val maxQueueSize: Int
    val loadingLimits: SparseArrayCompat<Float>
    val usedLoadingSpace: SparseArrayCompat<Int>
    val waitingTaskInfo: SparseArrayCompat<Int>
    val blockedTaskInfo: SparseArrayCompat<Int>

    fun startSnapshotRecording(taskManager: TaskManager)
    fun stopSnapshotRecording()

    fun addSnapshotListener(listener: OnSnapshotChangedListener)
    fun removeSnapshotListener(listener: OnSnapshotChangedListener)

    interface OnSnapshotChangedListener {
        fun onSnapshotChanged(snapshot: TaskManagerSnapshot)
    }
}
